Item SaveContext Logic

Malkierian opened the pull request:
This is part 1 of SaveContext Logic. All randomized items now set to and check against their respective locations on SaveContext, whether that be eqiupment, inventory, or flags. This also removes LogicVar pointers from the ItemTable and converts them into LogicVal enum references that are added to an instance-based `inLogic` vector instead that `UpdateHelpers` checks. This is only needed because of certain accessibility things, like BuySeed, FishAccess, etc, that are part of the ItemTable that the logic still needs to check against.

As a side note, the changes I made here make GetAccessibleLocations run *significantly* slower in debug configuration, but it's still just as fast as normal in release.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1767690904.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1767707971.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1767708011.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1767726476.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1767731892.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Port initial framework up to latest dev-rando. Includes about 15 items using the SCL structure for tracking.

Malkierian committed:
Most other item variables ported to SaveContext in UpdateHelpers.

Malkierian committed:
Removed individual Context retrievals in favor of single local variable in Logic to reference.

Malkierian committed:
start of some diagnostics.
caught some reverse logic with ApplyItemEffect.

Malkierian committed:
more diagnostics

Malkierian committed:
Resolve major logic issue with SetInventory flow in ApplyItemEffect.
Other small logic issues in ApplyItemEffect.

Malkierian committed:
Added a bit of logging.
Add Double Defense to item effects.

Malkierian committed:
revert hint gen bypass

Malkierian committed:
item_list *actually* divorced from logicVar pointers.
Several missed options for ApplyItemEffect, HasItem, and UpdateHelpers added in.
Changed HasItem to a switch block for more performance.
Seeds generate within 10-20 seconds in release mode. Further optimization is assumed possible.

Malkierian committed:
Some formatting

Malkierian committed:
Bit more formatting.

Pepe20129 commented on soh/soh/Enhancements/randomizer/logic.cpp:
```diff
@@ -340,15 +598,14 @@ namespace Rando {
         CanSummonGossipFairyWithoutSuns = CanUse(RG_ZELDAS_LULLABY) || CanUse(RG_EPONAS_SONG) || CanUse(RG_SONG_OF_TIME);
         CanSummonGossipFairy            = CanSummonGossipFairyWithoutSuns || CanUse(RG_SUNS_SONG);
-        Hearts          = BaseHearts + HeartContainer + (PieceOfHeart >> 2);
+        Hearts          = Hearts = ctx->GetSaveContext()->healthCapacity / 16;
```
```suggestion
        Hearts          = ctx->GetSaveContext()->healthCapacity / 16;
```

Pepe20129 commented on soh/soh/Enhancements/randomizer/logic.cpp:
```diff
@@ -619,38 +894,39 @@ namespace Rando {
-        OcarinaCUpButton = ctx->GetOption(RSK_SHUFFLE_OCARINA_BUTTONS).Is(true) ? 0 : 1;
-        OcarinaCDownButton = ctx->GetOption(RSK_SHUFFLE_OCARINA_BUTTONS).Is(true) ? 0 : 1;
+        bool ocBtnShuffle = ctx->GetOption(RSK_SHUFFLE_OCARINA_BUTTONS).Is(true);
+        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_A, ocBtnShuffle);
+        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_C_UP, ocBtnShuffle);
+        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_C_DOWN, ocBtnShuffle);
+        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_C_LEFT, ocBtnShuffle);
+        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_C_RIGHT, ocBtnShuffle);
```
```suggestion
        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_A, !ocBtnShuffle);
        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_C_UP, !ocBtnShuffle);
        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_C_DOWN, !ocBtnShuffle);
        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_C_LEFT, !ocBtnShuffle);
        ctx->SetRandoInf(RAND_INF_HAS_OCARINA_C_RIGHT, !ocBtnShuffle);
```

Malkierian replied:
No, this is correct, because SetRandoInf's bool is `remove`, so if button shuffle is on you want them "removed" (though they're just never added in the end).

Pepe20129 commented on soh/soh/Enhancements/randomizer/logic.cpp:
```diff
@@ -619,38 +894,39 @@ namespace Rando {
+        ProgressiveWallet = 0;
+        if (ctx->GetOption(RSK_SHUFFLE_CHILD_WALLET).Is(false)) {
+            ProgressiveWallet = 1;
+            ctx->SetRandoInf(RAND_INF_HAS_WALLET, false);
```
```suggestion
            ctx->SetRandoInf(RAND_INF_HAS_WALLET, true);
```

Malkierian replied:
See other comments about inversion.

Pepe20129 commented on soh/soh/Enhancements/randomizer/logic.cpp:
```diff
@@ -619,38 +894,39 @@ namespace Rando {
+        ProgressiveScale = 0;
+        if (ctx->GetOption(RSK_SHUFFLE_SWIM).Is(false)) {
+            ProgressiveScale = 1;
+            ctx->SetRandoInf(RAND_INF_CAN_SWIM, false);
```
```suggestion
            ctx->SetRandoInf(RAND_INF_CAN_SWIM, true);
```

Malkierian replied:
Also correct because of the inversion. I know it's potentially confusing, but I tried changing it to be affirmative and ran into problems.

Malkierian replied:
If it's desired for me to work out the affirmative form before merging, then we can talk about that, the inversion was just already well embedded into the way things were being done by the time I thought of trying to do it the other way, and proved complicated undoing.

Pepe20129 replied:
I'd say to merge and then fix later but yeah, I was confused.

Pepe20129 commented on soh/soh/Enhancements/randomizer/logic.cpp:
```diff
@@ -619,38 +894,39 @@ namespace Rando {
-        GerudoTrainingGroundsKeys = 0;
-        BottomOfTheWellKeys       = 0;
-        TreasureGameKeys          = 0;
+        if (IsKeysanity || ctx->GetDungeon(Rando::FIRE_TEMPLE)->IsMQ()) {
```
```suggestion
        if (!IsKeysanity && ctx->GetDungeon(Rando::FIRE_TEMPLE)->IsVanilla()) {
```

Pepe20129 commented on soh/soh/Enhancements/randomizer/logic.cpp:
```diff
@@ -124,37 +280,62 @@ namespace Rando {
+                return IsChild;
+
+                // Adult Trade
+            case RG_POCKET_EGG:
+                return IsAdult;
+            case RG_COJIRO:
+                return IsAdult;
+            case RG_ODD_MUSHROOM:
+                return IsAdult;
+            case RG_ODD_POTION:
+                return IsAdult;
+            case RG_POACHERS_SAW:
+                return IsAdult;
+            case RG_BROKEN_SWORD:
+                return IsAdult;
+            case RG_PRESCRIPTION:
+                return IsAdult;
+            case RG_EYEBALL_FROG:
+                return IsAdult;
+            case RG_EYEDROPS:
+                return IsAdult;
+            case RG_CLAIM_CHECK:
+                return IsAdult;
```
```suggestion
            case RG_POCKET_EGG:
            case RG_COJIRO:
            case RG_ODD_MUSHROOM:
            case RG_ODD_POTION:
            case RG_POACHERS_SAW:
            case RG_BROKEN_SWORD:
            case RG_PRESCRIPTION:
            case RG_EYEBALL_FROG:
            case RG_EYEDROPS:
            case RG_CLAIM_CHECK:
                return IsAdult;
```

Malkierian replied:
I left these separate because I don't know what the implications are for glitched play. Pretty sure at least a couple of them can be turned in as child with glitches.

Pepe20129 reviewed (commented):
> This is only needed because of certain accessibility things, like BuySeed, FishAccess, etc, that are part of the ItemTable that the logic still needs to check against.

I think it'd be better to just special case those by having more logic-only `RG`s or some other way to store the information. Primarily I think that keeping the whole system for a few cases is a bad idea.

Malkierian commented:
> I think it'd be better to just special case those by having more logic-only `RG`s or some other way to store the information. Primarily I think that keeping the whole system for a few cases is a bad idea.

I left them separate like that because of the possibility of adding EventAccess variables to the structure, and *something* was needed to replace the way shop rando referenced the logic vars (vector of them which were added as necessary for an ignore) which can't be handled just by the remaining LogicVars or special RGs as far as I can tell, and didn't match up exactly with any other field already in the Item class.

Malkierian committed:
Couple review changes.

Pepe20129 commented:
In that case I think all items that don't use this should have the none value instead of a specific one for each item as that implies there's differences between them.

Malkierian commented:
You should look at the discussion just had on Discord, there's more to it than just those access ones that I just didn't explain well.

Malkierian committed:
Merge branch 'develop-rando' of github.com:Malkierian/Shipwright into scl-redo

Malkierian committed:
Incorporate progressive upgrades and skeleton key into SCL.
Bit of cleanup.

Malkierian committed:
Change `inLogic` to static array.

Malkierian committed:
Merge branch 'develop-rando' of github.com:Malkierian/Shipwright into scl-redo

Malkierian committed:
Fix bug in `BottleCount()` preventing Jabu Jabu from ever being in logic.

Malkierian committed:
Fix bomb bag HasItem check.

Malkierian committed:
Merge branch 'scl-redo' of github.com:Malkierian/Shipwright into scl-redo

Malkierian committed:
Fix Gerudo Fortress and Ganon's Castle SmallKeys checks.

Malkierian committed:
Change all logic use of `bool remove` to be `bool state` and work on the affirmative assumption.

Malkierian committed:
Forgot the `Set(x)` uses in `Logic::Reset()`.

Malkierian committed:
Fix fishing pole effect application.

Malkierian committed:
fishing pole tweak

Malkierian committed:
Add `RG_BOMB_BAG`, `RG_NUTS` and `RG_STICKS` to `HasItem()`.
Fix silver and gold gauntlet checks in `HasItem()`.
Change Nuts and Sticks helpers to use non-progressive RGs in `UpdateHelpers()`.

Malkierian committed:
Change mSaveContext to simple pointer.
Set `mSaveContext` to `&gSaveContext` when loading a save file.
Modify `Item::GetGIEntry()` to use rando context's SaveContext pointer instead of the gSaveContext macros.
Utilize `Item::GetGIEntry()` to get real RG of infinite upgrades and set rando infs accordingly, skipping upgrade level incrementing (which was causing overflow of wallets for some reason).

Pepe20129 commented on soh/soh/Enhancements/randomizer/context.cpp:
```diff
@@ -395,6 +395,500 @@ void Context::ParseHintJson(nlohmann::json spoilerFileJson) {
     CreateStaticHints();
 }
 
+std::map<RandomizerGet, uint32_t> Context::RandoGetToFlag = {
+    { RG_KOKIRI_SWORD,           EQUIP_FLAG_SWORD_KOKIRI },
+    { RG_MASTER_SWORD,           EQUIP_FLAG_SWORD_MASTER },
+    { RG_BIGGORON_SWORD,         EQUIP_FLAG_SWORD_BGS },
+    { RG_DEKU_SHIELD,            EQUIP_FLAG_SHIELD_DEKU },
+    { RG_HYLIAN_SHIELD,          EQUIP_FLAG_SHIELD_HYLIAN },
+    { RG_MIRROR_SHIELD,          EQUIP_FLAG_SHIELD_MIRROR },
+    { RG_GORON_TUNIC,            EQUIP_FLAG_TUNIC_GORON },
+    { RG_ZORA_TUNIC,             EQUIP_FLAG_TUNIC_ZORA },
+    { RG_BUY_DEKU_SHIELD,        EQUIP_FLAG_SHIELD_DEKU },
+    { RG_BUY_HYLIAN_SHIELD,      EQUIP_FLAG_SHIELD_HYLIAN },
+    { RG_BUY_GORON_TUNIC,        EQUIP_FLAG_TUNIC_GORON },
+    { RG_BUY_ZORA_TUNIC,         EQUIP_FLAG_TUNIC_ZORA },
+    { RG_IRON_BOOTS,             EQUIP_FLAG_BOOTS_IRON },
+    { RG_HOVER_BOOTS,            EQUIP_FLAG_BOOTS_HOVER },
+    { RG_GOHMA_SOUL,             RAND_INF_GOHMA_SOUL },
+    { RG_KING_DODONGO_SOUL,      RAND_INF_KING_DODONGO_SOUL },
+    { RG_BARINADE_SOUL,          RAND_INF_BARINADE_SOUL },
+    { RG_PHANTOM_GANON_SOUL,     RAND_INF_PHANTOM_GANON_SOUL },
+    { RG_VOLVAGIA_SOUL,          RAND_INF_VOLVAGIA_SOUL },
+    { RG_MORPHA_SOUL,            RAND_INF_MORPHA_SOUL },
+    { RG_BONGO_BONGO_SOUL,       RAND_INF_BONGO_BONGO_SOUL },
+    { RG_TWINROVA_SOUL,          RAND_INF_TWINROVA_SOUL },
+    { RG_GANON_SOUL,             RAND_INF_GANON_SOUL },
+    { RG_OCARINA_A_BUTTON,       RAND_INF_HAS_OCARINA_A },
+    { RG_OCARINA_C_UP_BUTTON,    RAND_INF_HAS_OCARINA_C_UP },
+    { RG_OCARINA_C_DOWN_BUTTON,  RAND_INF_HAS_OCARINA_C_DOWN },
+    { RG_OCARINA_C_LEFT_BUTTON,  RAND_INF_HAS_OCARINA_C_LEFT },
+    { RG_OCARINA_C_RIGHT_BUTTON, RAND_INF_HAS_OCARINA_C_RIGHT },
+    { RG_SKELETON_KEY,           RAND_INF_HAS_SKELETON_KEY },
+    { RG_GREG_RUPEE,             RAND_INF_GREG_FOUND },
+    { RG_FISHING_POLE,           RAND_INF_FISHING_POLE_FOUND }
+};
```
A bit weird to have two types of flags in here.

Malkierian replied:
It works because RG is still unique in this, and allows it to be used for multiple case types at the same time, but if it's too confusing, I could separate them into two.

Pepe20129 commented on soh/soh/Enhancements/randomizer/item.cpp:
```diff
@@ -100,7 +67,7 @@ ItemType Item::GetItemType() const {
 }
 
-std::variant<bool*, uint8_t*> Item::GetLogicVar() const {
+LogicVal Item::GetLogicVar() const {
```
Due to the change, there's now `LogicVal` & `LogicVar`, probably a good idea to rename one to the other.

Malkierian replied:
Good catch.

Pepe20129 reviewed (approved):
Looks good! Two non-blocking things.
Also, maybe it'd be a good idea to use the new item functions on `gSaveContext` when getting an item during play instead of also having `Randomizer_Item_Give` but that's out of scope of this PR.

Malkierian committed:
Rename all remaining references of LogicVar to variants of LogicVal.

Malkierian committed:
Change final references to song LogicVars to CanUse calls, and remove them.

Malkierian committed:
Separate equip and rand inf flag maps.
Add Zelda's Letter and Weird Egg rand infs to map.
Consolidate `HasItem` rand inf  lookups using rand inf map and equips using equip flag map.
Maps are static, so can be referenced with `Context::` instead of ctx pointer.
